rfcs/20200113-tf-data-service.md

```diff
@@ -0,0 +1,591 @@
+valuable accelerator resources. The tf.data service solves this problem by using
+N input workers to feed M accelerators. The number of input workers can be
+scaled up or down as needed to keep up with the accelerators.
+
+### Distributed training requires a distribution-aware input pipeline.
+
+Today tf.data supports the tf.distribute API by providing mechanisms for
+sharding, cloning, and re-batching. The tf.distribute API uses these primitives
+to implement their own version of a distributed dataset. If distributed datasets
+become a core feature of tf.data, tf.data can provide a public API for
+tf.distribute (and users who wish to implement their own distribution) to use
+instead. This will also allow us to support feature requests that require
+cross-worker coordination, such as dynamic sharding.
+
+## User Benefit
+
+### Input-bound models
+
+Users with input-bound models can leverage the tf.data service to distribute
+input processing across horizontally-scaling compute resources. This can improve
+utilization for valuable accelerator resources, reducing total cost.
+
+### Dynamic load balancing
+
+Today, the tf.distribute API statically shards data across accelerators. This
+can lead to suboptimal utilization because some shards may contain more data
+than others. The tf.data service provides a mechanism for dynamically sharding,
+reducing the data imbalance across accelerators.
+
+### Visitation guarantees
+
+Model accuracy can often be improved when each training sample is trained on
+exactly once per epoch. The tf.data service can coordinate across workers to
+provide this guarantee.
+
+## Design Proposal
+
+The tf.data service is a master-worker system which iterates through datasets,
+producing outputs to be consumed by accelerators. The service is comprised of a
+few components:
+
+*   User-facing Python API for interacting with the tf.data service.
+*   Dataset splitting API for determining how to split up datasets for parallel
+    processing.
+*   Master and worker gRPC services.
+
+### Architecture
+
+The tf.data service is comprised of master and worker gRPC services which could
+be run in a couple of different configurations:
+
+#### Glossary
+
+**Master**: The single master coordinating the tf.data service.
+
+**Worker**: A tf.data service worker which performs dataset processing and
+provides dataset elements to consumers over RPC.
+
+**Consumer**: A machine which consumes data from the tf.data service. The
+consumer may be attached to a GPU or TPU, or use data for on-CPU training.
+
+#### Separate Cluster Architecture
+
+Each server is run on a separate host from the TensorFlow cluster. This
+configuration gives users a way to provide horizontally scaling CPU for
+processing their input pipelines and quickly feeding data to accelerators.
+
+#### Embedded Cluster Architecture
+
+without needing to provision additional compute resources. and gives all the
```

Comment:
Weird sentence structure; was this an editing mistake?

Reply:
Yeah, the previous two lines were deleted by accident. Fixed now, thanks